[TG-1121] Bugfix for multi-dimensional arrays with non-const size and non-const access

majakusber opened the pull request:
This resolves the bug when `cannot unpack array of non-const size` crash occurred for multi-dimensional arrays with non-const size and non-const array access. The problem was that during symex the array access was attempted using `byte_extract` instruction because the subtype of the multi-dimensional array got lost during dereferencing.

The solution allows any pointer to be dereferenced to void pointer (during symex).

Commit:
Allow pointers to be dereferenced to void types

Commit:
Enable previously failing regression tests

Commit:
Update regression test that no longer throws an exception

thk123 commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -202,7 +202,12 @@ bool value_set_dereferencet::dereference_type_compare(
+    return true; // always ok (anything can be dereferenced to void type)
+
+  if(
+    object_type.id() == ID_pointer && dereference_type.id() == ID_pointer &&
```
The first if handles `void*`, this one handles `void**` - what about `void***`? (not possible I assume in Java but perhaps in C?)

tautschnig replied:
Is it necessary to restrict this to a matching number of `ID_pointer` levels? So `int**` vs. `void**` is ok, but `int*` vs. `void**` isn't?

majakusber replied:
Good points. I took a shot at it in the next commit, let me know what you think.

thk123 reviewed (commented):
LGTM - would obviously be best to have a test to explores this `void***` and mismatched depth scenarios, presumably in `C`, do either of @peterschrammel or @tautschnig have any suggestions of the kind of code that might exhibit this?

thk123 commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -201,8 +201,25 @@ bool value_set_dereferencet::dereference_type_compare(
+  if(
+    dereference_unwrapped.id() == ID_empty &&
+    object_unwrapped.id() != ID_pointer)
+    return true;
```
Nit: since the condition of the if is multi line I suggest wrapping this `return true` in braces to make it clearer.

tautschnig commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -201,13 +201,25 @@ bool value_set_dereferencet::dereference_type_compare(
+  // - object_type=(int **), dereference_type=(void **) is ok;
+  // - object_type=(int *), dereference_type=(void **) is not ok;
+  typet object_unwrapped = object_type;
+  typet dereference_unwrapped = dereference_type;
```
I would have used pointers here, but I won't insist. There are a couple of places in the codebase that dig through array or pointer types, and those tend to use pointers to avoid repeated construction of a non-trivial object.

peterschrammel replied:
I tend to insist that pointers should be used here.

tautschnig commented:
> would obviously be best to have a test to explores this `void***` and mismatched depth scenarios, presumably in C, do either of @peterschrammel or @tautschnig have any suggestions of the kind of code that might exhibit this?

I think that something like `int a[1][1]; int **a_ptr = a; void *v_ptr = a_ptr; int x = (int *)v_ptr;` might trigger this? Testing and debugging would be really appreciated, if feasible! I believe that in many cases the effect of this PR will only be noticed through improved performance, and thus might go completely unnoticed.

peterschrammel commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -201,13 +201,25 @@ bool value_set_dereferencet::dereference_type_compare(
+    dereference_unwrapped = dereference_unwrapped.subtype();
+  }
   if(
-    object_type.id() == ID_pointer && dereference_type.id() == ID_pointer &&
```
Please squash this commit with the first one.

peterschrammel commented:
> I believe that in many cases the effect of this PR will only be noticed through improved performance, and thus might go completely unnoticed.

The problem that triggered this bugfix are arrays of polymorphic types in java, where goto-symex failed to narrow to the actual type when indexing an array of void pointers, introducing byte extract expressions that cannot be flattened.

Commit:
Adding debug information to dereference type comparison

majakusber commented:
I updated to code according to reviews, added additional case to the type comparison and some debug information. 

For testing this - we have several regression tests that demonstrate that things now work as expected for arrays (see the commits). I was trying to add an example that would go wrong, but didn't succeed. The example suggested by @tautschnig does not trigger this. It only gives a warning during goto-program initialisation during typecheck, the warning being `incompatible pointer types`. Since the `dereference_type_compare` function is private, it's not possible to write isolated unit tests for it without changing the class itself and for the public function that calls it it's also too complex.

Let me know if you would agree with merging this with successful tests only or if you think a failing example is needed too.

tautschnig commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -201,8 +201,27 @@ bool value_set_dereferencet::dereference_type_compare(
+    dereference_unwrapped = &dereference_unwrapped->subtype();
+  }
+  if(
+    dereference_unwrapped->id() == ID_empty)
```
Nit pick: this line breaking seems unnecessary now.

tautschnig commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -222,6 +224,19 @@ bool value_set_dereferencet::dereference_type_compare(
+                    << messaget::endl;
+    message.debug() << " dereference_type: " << dereference_type.pretty()
+                    << messaget::eom;
+  }
```
I don't think this debug output should stay in, because it newly introduces a message handler that might not match the front-end's one.

tautschnig replied:
Actually the entire commit "Adding debug information to dereference type comparison" should be dropped, I believe.

majakusber replied:
I was unsure whether this 'forced' debug is a good idea actually. The thing is that it took us a long time to find the problem in symex so for future cases this output could save us time. But I agree it's not ideal.

tautschnig replied:
"forced debug" is better known as "warning" :-)  So if you want to keep it in (and you have given good reasons for doing so), then please make sure there is a message handler in place so that proper output can be generated. If that's a bigger project than is feasible right now, you might just create a fresh PR with just this commit in it?

majakusber replied:
I meant forced in that I create a message handler especially for this :)

tautschnig replied:
I'm not sure I understand; depending on who should be able to control this output it would either be guarded by `#ifdef DEBUG` or be passed a configured message handler.

majakusber replied:
I wanted to do it via a message handler (and changing the `debug()` above into `warning()`) but passing the message handler is a nightmare... so I did it as `#ifdef DEBUG` instead.

majakusber commented:
There is a failing test on the test-gen PR, connected to enums. I added a 'do not merge' label while I investigate it.

thk123 commented:
Rebased to see if subsequent fixes fix the bump for this to.

thk123 reviewed (approved):
TG pointer bump passes.
